caas/kubernetes/provider/application/constraints.go

```diff
@@ -266,35 +315,139 @@ func processPodAffinity(pod *core.PodSpec, affinityLabels map[string]string) err
 	return nil
 }
 
-func configureConstraint(pod *core.PodSpec, resourceName core.ResourceName, value string) (err error) {
+// divideAndSpreadContainerResource evenly distributes the total resource across a given number of containers.
+// It returns a slice of len(containers), where each element represents the distributed value
+// allocated to that container. Any remainder is distributed one-by-one starting from the lowest index.
+// e.g., [4,4,3] for totalResource = 11, numContainers = 3
+func divideAndSpreadContainerResource(totalResource uint64, numContainers int) []uint64 {
+	if numContainers <= 0 {
+		return nil
+	}
+
+	result := make([]uint64, numContainers)
+	quotient := totalResource / uint64(numContainers)
+	remainder := totalResource % uint64(numContainers)
+
+	for i := 0; i < numContainers; i++ {
+		if uint64(i) < remainder {
+			result[i] = quotient + 1
+		} else {
+			result[i] = quotient
+		}
+	}
+	return result
+}
+
+// configureWorkloadConstraint sets the resource limits for all containers,
+// and the resource request for the first workload container only.
+func configureWorkloadConstraint(pod *core.PodSpec, resourceName core.ResourceName, value uint64) (err error) {
 	if len(pod.Containers) == 0 {
 		return nil
 	}
-	pod.Containers[0].Resources.Requests, err = MergeConstraint(resourceName, value, pod.Containers[0].Resources.Requests)
+	var unitSuffix string
+
+	switch resourceName {
+	case core.ResourceMemory:
+		unitSuffix = "Mi"
+	case core.ResourceCPU:
+		unitSuffix = "m"
+	default:
+		return errors.NotSupportedf("converting resource value for %q", resourceName)
+	}
+
+	resourceValStr := fmt.Sprintf("%d%s", value, unitSuffix)
+	resourceQty, err := resource.ParseQuantity(resourceValStr)
 	if err != nil {
-		return errors.Annotatef(err, "merging request constraint %s=%s", resourceName, value)
+		return errors.Annotatef(err, "parsing constraint value %q for %q", value, resourceName)
+	}
+	pod.Containers[0].Resources.Requests, err = MergeConstraint(resourceName, resourceQty, pod.Containers[0].Resources.Requests)
```

Comment:
How do you know that there is at least 1 container in the pod?

Reply:
Hi tom, this is the current implementation (if we look at deleted line 273) which caused the bug where statefulset does not match the actual resource applied, by right we shouldn't be touching this. I just refactored this line to somewhere later because I changed the signature of value to make it more consistent with the new V2 implementation as well as allowing extension of resourceName to be easier should we need to introduce them in the future for the prev versions. I have also checked with Ian on this that a charm container should always be present in a charm. However, I have added a quick check on the len of containers before setting the reqs to improve on the defensive prog for the current impl.